### PR TITLE
torch mlmodel

### PR DESCRIPTION
**Pull request by Rob1in**
# Open questions
## name of the module:

1. I think GPU support is **very** straightforward. It's [there](https://github.com/Rob1in/torch/blob/main/src/torch_mlmodel_module.py#L68) and I just need to test it on an Orin I guess. it works on the `maingear` so the name of the module is `torch` and not `torch-cpu`. It's not on the registry yet, so let me know if I should change it?

## `infer()`
2.  Are single input and output tensors ok for a first version? I haven't investigated yet how does a multiple input/output tensors model look like in PyTorch? I explained what the keys of the input/output `Dict`s should be [here](https://github.com/Rob1in/torch/blob/main/src/torch_mlmodel_module.py#L68). 

 ## `metadata()`
 - I haven't figured out the  `type` field in `TensorInfo` yet. `'float32'` hardcoded for now. 
 
 # Follow-up PRs
 1. PyINstaller
 2. GPU (or not)
 3. Support pertained torchvision models. Need to talk about how the config should look like.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,82 @@
...
+      "type": "local"
+    }
+  ],
+  "services": [
+    {
+      "name": "torch",
+      "type": "mlmodel",
+      "model": "viam:mlmodel:torch",
+      "attributes": {
+        "path_to_serialized_file": "examples/resnet_18/resnet-18.pt",
```
Comment (bhaney) [inline README.md]: this should just be called "model_path"

File: README.md [hunk 2]
```diff
@@ -1 +1,82 @@
...
+
+
+# Methods
+## `infer()`
+```
+infer(input_tensors: Dict[str, NDArray], *, timeout: Optional[float]) -> Dict[str, NDArray]
+```
+
+### Input and output dictionnaries.
+ - For now, only support single input tensor. Key in the input `Dict` must be `'input'`.
```
Comment (bhaney) [inline README.md]: if you only require a single input tensor, the key doesn't matter, and it shouldn't need to be a requirement

File: README.md [hunk 3]
```diff
@@ -1 +1,82 @@
...
+# Methods
+## `infer()`
+```
+infer(input_tensors: Dict[str, NDArray], *, timeout: Optional[float]) -> Dict[str, NDArray]
+```
+
+### Input and output dictionnaries.
+ - For now, only support single input tensor. Key in the input `Dict` must be `'input'`. 
+- For now, only support single output tensor. 
+Key in the output `Dict` is `'output'`.
```
Comment (bhaney) [inline README.md]: I do think that this should be changed to support multiple output tensors, and it should be `output_X`

File: README.md [hunk 4]
```diff
@@ -1 +1,82 @@
...
+
+
+### Attributes description
+
+The following attributes are available to configure your module:
+
+
+| Name         | Type   | Inclusion    | Default | Description                       |
+| ------------ | ------ | ------------ | ------- | --------------------------------- |
+| `model_file` | string | **Required** |         | Path to **standalone** model file |
```
Comment (bhaney) [inline README.md]: You'll also wanted to include "label_path" and put the label file name into the "Extra" field of the metadata for each output tensors. You can see how it's done in the onnx-cpu model, or also Khari's tensorflow model

File: requirements.txt
Comment (bhaney) [inline requirements.txt:1]: Please freeze the versions in all of these dependencies

File: src/model/model.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+class TorchModel:
+    def __init__(
+        self, path_to_serialized_file: str, path_to_model_file: str = None
+    ) -> None:
+        self.model = torch.load(path_to_serialized_file)
+        if not isinstance(self.model, nn.Module):
+            if isinstance(self.model, OrderedDict):
+                LOGGER.error(
+                    f"the file {path_to_model_file} provided as model file is of type collections.OrderedDict, which suggests that the provided file describes weights instead of a standalone model"
```
Comment (bhaney) [inline src/model/model.py]: I love the detail on this error, well written!

File: src/torch_mlmodel_module.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+from viam.logging import getLogger
+from .model.model import TorchModel
+from .model_inspector.inspector import Inspector
+import torch
+
+LOGGER = getLogger(__name__)
+
+
+class TorchMLModelModule(MLModel, Reconfigurable):
+    MODEL: ClassVar[Model] = Model(ModelFamily("viam", "mlmodel"), "torch")
```
Comment (bhaney) [inline src/torch_mlmodel_module.py]: You may want to separate this out into another file because youll want to eventually make two models: torch-cpu and torch-gpu. Given that almost nothing will change between the two modules, make it easy to grab the model name from an external function and place it in here

File: src/torch_mlmodel_module.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+        self.path_to_model_file = get_attribute_from_config("model_file", None, str)
+        self.path_to_label_file = get_attribute_from_config("label_file", None, str)
+        self.model_type = get_attribute_from_config("model_type", None, str)
+
+        self.torch_model = TorchModel(path_to_serialized_file=self.path_to_model_file)
+        self.inspector = Inspector(self.torch_model.model)
+        self.input_shape, self.output_shape = self.inspector.find_metadata()
+        self.input_names = ["input"]
+        self.output_names = ["output"]
```
Comment (bhaney) [inline src/torch_mlmodel_module.py]: You can make input and output names the same length as the number of output tensors you found above

File: src/torch_mlmodel_module.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+        # TODO: Test self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+        self.path_to_model_file = get_attribute_from_config("model_file", None, str)
+        self.path_to_label_file = get_attribute_from_config("label_file", None, str)
+        self.model_type = get_attribute_from_config("model_type", None, str)
+
+        self.torch_model = TorchModel(path_to_serialized_file=self.path_to_model_file)
+        self.inspector = Inspector(self.torch_model.model)
+        self.input_shape, self.output_shape = self.inspector.find_metadata()
```
Comment (bhaney) [inline src/torch_mlmodel_module.py]: both output_shape and input_shape will have to be lists - it's an explicit scope goal that multiple inputs and outputs can be handled.

File: src/torch_mlmodel_module.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        Args:
+            input_tensors (Dict[str, NDArray]): A dictionary of input flat tensors as specified in the metadata
+
+        Returns:
+            Dict[str, NDArray]: A dictionary of output flat tensors as specified in the metadata
+        """
+        res = dict()
+        with torch.no_grad():
+            single_input = input_tensors[self.input_names[0]]
+            res[self.output_names[0]] = self.torch_model.infer(single_input)
```
Comment (bhaney) [inline src/torch_mlmodel_module.py]: you'll have to change this to be able to take multiple outputs

File: src/torch_mlmodel_module.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+        Returns:
+            Metadata: The metadata
+        """
+        input_infos = [
+            TensorInfo(name=input_name, data_type="float32", shape=self.input_shape)
+            for input_name in self.input_names
+        ]
+        output_infos = [
+            TensorInfo(name=output_name, data_type="float32", shape=self.output_shape)
```
Comment (bhaney) [inline src/torch_mlmodel_module.py]: You will need to put "extra" in each output tensor and put in the label_path if it is provided by the user

Comment (bhaney) [inline src/torch_mlmodel_module.py]: The data_type will not always be "float32", and you'll have to change the logic to make it take into account multiple kinds of output shapes

File: src/model_inspector/inspector.py
Comment (bhaney) [inline src/model_inspector/inspector.py:1]: Unfortunately, the logic for the inspector as it is written is extremely limited - it requires that the model only have one input and one output. This unfortunately cannot fufill the scope goal for multiple inputs/outputs in an easy way. I would say - start simply first. Make basic assumptions to allow for the inspector to return multiple input and output tensor. The way it is written too, I don't know if this can be easily extended to handle multiple inputs/outputs -- the way the children logic is written does not allow for branching capabilities.

**Review by bhaney — CHANGES_REQUESTED** [review]
A good first pass when it comes to inferring the shapes of the input tensor, but the logic is very strict in how it only allows one input and output tensor. You will need to re-write the inspector to be able to handle multiple input and outputs. It's better to be able to handle the most common types of models, and fail for weird edge cases, rather than making sure you can handle all edge cases of a restricted case (one input, and one output).

The scope goal specifically states the service must be able to handle multiple inputs and outputs. You cannot add this module to registry without these capabilities.

File: src/model/model.py [hunk 11]
```diff
@@ -1,31 +1,65 @@
...
+        elif isinstance(output, Iterable):
+            res = {}
+            count = 0
+            for out in output:
+                res[f"output_{count}"] = out
+                count += 1
+            return res
+
+        else:
+            raise TypeError()
```
Comment (bhaney) [inline src/model/model.py]: You will want to add details to this error

File: src/model_inspector/input_tester.py [hunk 12]
```diff
@@ -37,11 +40,11 @@ def dimensionality_unicity(sizes: List[List[int]]) -> Optional[int]:
             Optional[int]: The dimensionality if it is unique, otherwise None.
```
Comment (bhaney) [inline src/model_inspector/input_tester.py]: Update this comment to reflect the new output type

File: src/model_inspector/input_tester.py [hunk 13]
```diff
@@ -106,32 +109,51 @@ def try_audio_input(self):
...
-        else:
-            self.not_working_input_sizes.append(input_size)
+    def test_input_size(self, input_size, add_batch_dimension=False):
+        input_array = torch.ones(
+            (input_size)
+        ).numpy()  # i get type issues when using np.ones()
+        input_tensor = {"input": input_array}
+        output = None
+        if add_batch_dimension:
+            input_tensor.unsqueeze(0)
```
Comment (bhaney) [inline src/model_inspector/input_tester.py]: the input_tensor is a dictionary, I think you mean to apply this on the value for the "input" entry. This will panic as its written

**Review by bhaney — CHANGES_REQUESTED** [review]
Nice work - Just a few small changes, mostly around adding comments, and one change to prevent a panic from happening.

**Comment by bhaney** [comment]
And don't forget about renaming the module to torch-cpu

**Review by bhaney — APPROVED** [review]
LGTM
